[WIP] concepts: move from previous docs to new section, etc.

jeremydesroches opened the pull request:
UPDATE: Latest status at https://github.com/iterative/dvc.org/pull/1944#issuecomment-739065804 (at the time of writing)

Per #550, creates a new **Basic Concepts** section in the user guide.

> Initially an outline of extracted descriptions from requested topics/tooltips for scoping: single document vs. multiple and depth of description. 

Summary of goals and plans from https://github.com/iterative/dvc.org/issues/550#issuecomment-673585050: 
 
> * [x]  There should be a page or section (maybe called **Basic Concepts** in UG, see example in [this closed PR](https://github.com/iterative/dvc.org/pull/1655/files?file-filters%5B%5D=#diff-cfffe134a2be45efa832d561bda45797)) — properly written on a user-level language (closer to UC) — explaining some concepts and DVC philosophy
>   > concepts like Cache, Remote, Wokspace, etc, and their relationships (with diagrams if needed)
> * [x] Possibly extract concept explanations from cmd refs (e.g. cache, dag, metrics/plots, remote)
  > * [x]  Figure out a connection to glossary tooltips (currently special md files with Frontmatter in `content/docs/user-guide/basic-concepts/` — see #1395): Just link to the Concepts? Use a similar engine (with Frontmatter)? Merge completely?
  > * [ ] Find other issues that would be included in solving this (e.g. #53, #1127).
> * ~~Consider also a related DVC Principles section (more philosophical, maybe in the Concepts section index)? See [#550 (comment)](https://github.com/iterative/dvc.org/issues/550#issuecomment-544106096) above.~~ Too much for this PR

**Concepts to be included**

- [x] Workspace
- (Versioning through) Codification ?
- [x] DVC Metafiles (may be part of ^)
- [x] DVC Cache
- [x] Data Pipelines
- [x] Remote Storage
- [x] Metrics and Plots
- [x] Parameters

---

**Concepts without main place in existing docs**? (for a future PR)

- DVC Project (glossary entry exists, may be combined with workspace)

Other notes:

- [ ] rename DVC Dirs&Files guide and DVC metafiles concept page later per https://github.com/iterative/dvc.org/pull/1944#discussion_r533822230
  > But don't extract detailed spec from guide (just the concept explanation)! It will be a separate task to figure that part out.

jeremydesroches commented:
@jorgeorpinel created this to address #550. Just an initial outline so far but feel free to comment on progress as I go. I'll request your review when decisions required or closer to merge, thanks.

shcheklein commented:
@jeremydesroches @jorgeorpinel qq- how will it replace/augment already existing basic concepts? I see the intention (one single page vs many pages), but we'll have to duplicate pretty much the information and support it later.

jeremydesroches commented:
> how will it replace/augment already existing basic concepts? I see the intention (one single page vs many pages), but we'll have to duplicate pretty much the information and support it later.

I think that the existing basic concepts, which are set up like a glossary, could remain and Basic Concepts would expand on select concepts. 

The glossary approach is quite useful and it would be odd to exclude the "concepts". Instead, I suggest linking to the concept from the glossary items that have them. 

For example, the dvc-cache.md glossary/tooltip links to DVC Files and Directories, so we might extract that to the Basic Concepts doc and apply a parallel structure to the other concepts.

With the above approach you could continue building out the glossary while also expanding on the more conceptual items. It avoids having to decide if a given term is best in the glossary or concepts — it can be both. 

In practice all "concepts" should probably have a glossary entry, but not all glossary terms are concepts.

jorgeorpinel commented:
> qq- how will it replace/augment already existing basic concepts?

Meaning the glossary/tooltips @shcheklein ? Yes. That is one of the questions:

> - [ ] Figure out a connection to glossary tooltips (currently special md files with Frontmatter in `content/docs/user-guide/basic-concepts/`

Not sure... But the tooltips are means as very short summaries so we may need to maintain both? The longer explanations can grow considerably, e.g. DVC remotes as currently explained in https://dvc.org/doc/command-reference/remote.

> one single page vs many pages

In fact we're not very sure of this format. May need one page per concept both because some concepts will be long to explain, and for SEO purposes (see https://github.com/iterative/dvc.org/issues/550#issuecomment-726434615).

jorgeorpinel commented:
> existing basic concepts, which are set up like a glossary, could remain

@jeremydesroches I would rename that folder to `content/docs/user-guide/glossary/` though, BTW — but there's an issue about that: #1395, we may just want to reactivate it (let's see what Ivan thinks).

Agree about linking from tooltips to concepts.

> not all glossary terms are concepts

➕

jorgeorpinel commented on content/docs/user-guide/basic-concepts.md:
```diff
@@ -0,0 +1,129 @@
+# Basic Concepts
+
+Intro and DVC philosophy...possible diagram of cache/remote/workspace
```
Well, the most basic concept in DVC is that of versioning through codification (currently being worked in #1747). So we could mention that and link to the Versioning use case as an intro.

jorgeorpinel replied:
I see this page no longer exists so will resolve. We could consider to use an index page here to address:

> - Consider also a related DVC Principles section (more philosophical, maybe in the Concepts section index)? See https://github.com/iterative/dvc.org/issues/550#issuecomment-544106096 above.

But for now I've scratched out that point from the goals of this PR to improve the scope.

jorgeorpinel commented on content/docs/user-guide/basic-concepts.md:
```diff
@@ -0,0 +1,129 @@
+
+Intro and DVC philosophy...possible diagram of cache/remote/workspace
+
+## Cache
+
+_From `dvc cache`_
```
Thanks for these notes. Could be `<!-- HTML comments -->` for now if you want.

jorgeorpinel replied:
Or actually, if you can list in the description of the PR the list of initial concepts you intend to put here and where they exist currently, that would be idea.

jorgeorpinel replied:
p.s. If there are any that you think should also be included but don't have a main place in existing docs let's list them too but we can address them in a future PR. This one should be mostly about concentrating existing content to the new doc/section, and link whatever is needed (for SEO purposes).

jeremydesroches replied:
I browsed and thought on this for a bit and didn't find any obvious candidates. Listed Project as a possible opportunity. `params` is a good one, but it's already in the glossary (and now as a concept.)

jorgeorpinel commented on content/docs/user-guide/basic-concepts.md:
```diff
@@ -0,0 +1,129 @@
+Learn more about it's
+[structure](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory).
+
+_from tooltip_
+
+The DVC cache is a hidden storage (by default located in the `.dvc/cache`
+directory) for files that are tracked by DVC, and their different versions.
+Learn more about it's
+[structure](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory).
```
OK I take it it's an intermediate step to put this here right? Because we don't want to duplicate these texts exactly, as the tooltips will (most likely) also remain.

jeremydesroches replied:
> OK I take it it's an intermediate step to put this here right?

Yes, I placed it as a mock-up for discussion purposes.

jorgeorpinel replied:
Cool

shcheklein commented:
Okay, to share a bit of context they way I see it. Existing "basic concepts" framework was made in way to accommodate longer texts (that's the actual basic concepts thingy) + glossary (that comes from the short description that is provided via frontmatter) + it can drive SEO (again via frontmatter). 

Make take from this discussion is that we should be doing something else only if we want to have everything on the single page.

> In practice all "concepts" should probably have a glossary entry, but not all glossary terms are concepts.

this is true, but it doesn't break the existing framework in any way, right?

jeremydesroches commented:
> Existing "basic concepts" framework was made in way to accommodate longer texts (that's the actual basic concepts thingy) + glossary (that comes from the short description that is provided via frontmatter) + it can drive SEO (again via frontmatter).

Thanks @shcheklein, this context helps a lot. 

> Make take from this discussion is that we should be doing something else only if we want to have everything on the single page.

Given the above context thats how it seems to me, too. To me this is a win-win — I also like the frontmatter approach to create a tooltip, while still having underlying (expanded) content for clickthrough or to drive SEO. 

That would mean we're taking the multiple document approach... sound OK @jorgeorpinel ?

> > In practice all "concepts" should probably have a glossary entry, but not all glossary terms are concepts.
> 
> this is true, but it doesn't break the existing framework in any way, right?

Right, we could add the concepts pages (or page) using the existing framework.

jorgeorpinel commented:
> Existing "basic concepts" framework was made to accommodate longer texts + glossary (short description via frontmatter).

Oh, I didn't realize md files in `content/docs/user-guide/basic-concepts/` support a description in Frontmatter. Currently they only use the `name` and `match` fields. If that's the case then let's definitely just edit them there, but we'll need a separate ticket to implement displaying them as pages of a special Basic Concepts section with the engine.

> we're taking the multiple document approach... sound OK?

Yes @jeremydesroches. But the most important thing you can provide for starters is the content, whatever the final structure is. Please focus on that in parallel with this discussion. I feel like we still need to catch up on progress on that front.

Thanks

jeremydesroches commented:
> Oh, I didn't realize md files in `content/docs/user-guide/basic-concepts/` support a description in Frontmatter. Currently they only use the `name` and `match` fields. If that's the case then let's definitely just edit them there, but we'll need a separate ticket to implement displaying them as pages of a special Basic Concepts section with the engine.

I might be missing something but I don't follow how the description frontmatter would work and what the advantage is. That is, using the current glossary/tooltip form to link to Basic Concepts, either as a single document with anchors or multiple documents within a section. I tried adding a `description:` block to test and saw no change to the tooltip behaviour.

In any case, I'll work on content for the concepts in parallel so we can place them in whatever structure is decided.

shcheklein commented:
> I tried adding a description: block to test and saw no change to the tooltip behaviour.

that can be fixed I think. @rogermparent can we make the engine take the text for tooltips from the frontmatter's description in this case?

@jeremydesroches the question is- should we use the same description in this case for SEO, or should we create two different fields? like `glossary` and `description`?

jeremydesroches commented:
> @jeremydesroches the question is- should we use the same description in this case for SEO, or should we create two different fields? like `glossary` and `description`?

Good question. The constraints for the meta description (less than 160 characters, doesn't like markup) would make it difficult to address both with the same content. For example, the existing descriptions/definitions have links and are longer than 160 characters.

Your suggestion would work. Even if there was no unique meta description, good answers to the query in the body copy will get picked up automatically.

rogermparent commented:
> that can be fixed I think. @rogermparent can we make the engine take the text for tooltips from the frontmatter's description in this case?

Yes, we can specifically parse a field as Markdown at build time and use its HTML output to provide glossary text.
If we decide to co-locate pure glossary entries and full Basic Concepts pages in `basic-concepts`, we can have glossary entries fall back to the body if that field is not present.

rogermparent commented:
I just made PR #1954 against this branch, it includes the ability to override glossary entries with a new frontmatter entry named `tooltip` that also supports Markdown like the current glossary entries.

jeremydesroches commented:
Hey @rogermparent, the tooltip markdown works great but it seems that the glossary pages are being skipped for rendering. So if I add them to the navigation with sidebar.json or link to them directly the slug is there but the page is rendered blank.

To mock up what (I think) we are after here, I added a new "Concepts" folder and broke out the topics on separate pages. This isn't the latest content but I wanted to check on the direction. 

My understanding of the desired behaviour is that glossary items could be added to the nav and rendered, appearing like these Concepts entries. The frontmatter on each would define the tooltip and the markdown would get rendered as a doc page. This would leave us with a single-source for each concept and/or glossary definition.

Does that sound right @jorgeorpinel and @shcheklein ?

shcheklein commented:
> My understanding of the desired behaviour is that glossary items could be added to the nav and rendered, appearing like these Concepts entries. The frontmatter on each would define the tooltip and the markdown would get rendered as a doc page. This would leave us with a single-source for each concept and/or glossary definition.

👍 yep, that's my understanding as well.

jorgeorpinel commented on content/docs/user-guide/concepts/dvc-files.md:
```diff
@@ -0,0 +1,26 @@
+# DVC Files
```
Not sure we need this in basic concepts. There's a guide for these. Also, the term is "DVC Metafiles" (referring to .dvc and dvc.yaml/lock files). But even extracting those, there's still DVC internal files & dirs that should stay in the existing guide.

So the question is should we/ is it worth it to move the metafiles out of the existing guide into here?

jeremydesroches replied:
> So the question is should we/ is it worth it to move the metafiles out of the existing guide into here?

Let's try to answer this after seeing how the other concepts shape up. Working on it.

jorgeorpinel commented on content/docs/sidebar.json:
```diff
@@ -87,6 +87,18 @@
+        "slug": "concepts",
+        "source": false,
+        "children": [
+          "cache",
+          "dvc-files",
+          "metrics-plots",
+          "pipelines",
+          "remote",
+          "workspace"
```
Ordered from more to less basic:
```suggestion
          "workspace",
          "dvc-metafiles",
          "cache",
          "pipelines",
          "metrics-plots",
          "remote"
```

jeremydesroches replied:
Thanks for the notes on order... I'll update these to follow the labels in the issue description.

jorgeorpinel replied:
K. And actually remote is probably more basic than metrics/plots.

jorgeorpinel commented on content/docs/user-guide/concepts/pipelines.md:
```diff
@@ -0,0 +1,20 @@
+# Pipelines
```
```suggestion
# Data Pipelines
```

jorgeorpinel commented on content/docs/user-guide/concepts/remote.md:
```diff
@@ -0,0 +1,17 @@
+# Remote
```
```suggestion
# Remote Storage
```

jorgeorpinel commented on content/docs/user-guide/glossary/dvc-project.md:
```diff
@@ -0,0 +1,28 @@
+`dvc.yaml` and `.dvc` files created with commands such as `dvc add` or
+`dvc run`.
+
+This page extends for a _while_, elaborating on what a DVC Project is in a way
+where it doesn't make sense to put **all** this stuff in a tooltip.
```
Where can we see this page in the review app (link pls)? Thanks

jeremydesroches replied:
This was example copy that Roger added to show up the new functionality, that was merged into my branch. DVC Project is a glossary item but not in our list for Basic Concepts so it's not in the nav and I'll revert it to the previous form.

jorgeorpinel replied:
Yep. This is no longer relevant. Resolving, thanks

jorgeorpinel commented on content/docs/user-guide/glossary/workspace.md:
```diff
@@ -6,3 +6,6 @@ match: [workspace]
 Directory containing all your project files e.g. raw datasets, source code, ML
 models, etc. Typically, it's also a Git repository. It will contain your DVC
 project.
+
+Learn more about the [workspace concept](/doc/user-guide/concepts/workspace) in
+DVC.
```
Try to stick to a single paragraph for tooltips, check out how they look in the review app. That last sentence for the link to basic concepts can probably be shorter too.

jorgeorpinel replied:
And I guess we're adding similar links to all tooltip texts?

jorgeorpinel replied:
But wait... I'm confused. I thought we were keeping a single file per concept with both the `glossary` text in Frontmatter, and the md content for the actual docs?

jeremydesroches replied:
Agreed on single paragraph. For the links I think it will come down to which items have their own page (basic concepts) and which ones remain as glossary items (same as previous behaviour). Yes, a single file per concept.

jorgeorpinel commented:
> > should we use the same description in this case for SEO, or should we create two different fields?
>
> difficult to address both with the same content

@shcheklein I agree with @jeremydesroches on this. The SEO description is (yet) another different text (with different goals and key words) we'd need to maintain, which won't usually be very similar to the tooltip text.

Question: do docs in basic-concepts md files support the `description` field in Frontmatter already?

I'm thinking maybe another alternative (instead of the current `glossary` field) is to only keep `description` (for SEO) and for the tooltips, extract just the first paragraph of the basic concept docs. WDYGT? (G=guys) cc @rogermparent

rogermparent commented:
> Hey @rogermparent, the tooltip markdown works great but it seems that the glossary pages are being skipped for rendering. So if I add them to the navigation with sidebar.json or link to them directly the slug is there but the page is rendered blank.

You're correct, pages in that glossary folder are skipped in doc page creation currently. The removal of one line will make them created just like regular docs pages again. I assumed we would want to do this at some point, so the line is even marked with a comment.

> To mock up what (I think) we are after here, I added a new "Concepts" folder and broke out the topics on separate pages. This isn't the latest content but I wanted to check on the direction.
> 
> My understanding of the desired behaviour is that glossary items could be added to the nav and rendered, appearing like these Concepts entries. The frontmatter on each would define the tooltip and the markdown would get rendered as a doc page. This would leave us with a single-source for each concept and/or glossary definition.

Totally doable! May I add a commit to re-enable glossary page creation?

jeremydesroches commented:
> Totally doable! May I add a commit to re-enable glossary page creation?

Yes please, @rogermparent. That way we can see exactly how this approach could work. Thanks!

rogermparent commented:
I've pushed some commits that that re-enable blog page creation and change sidebar.json to point to the glossary directory to show off docs pages in glossary. I'm certain this is a kludgy representation that will need to be changed, but it gets the idea across.

[Check it out here!](https://dvc-landing-ug-basic-co-kn0p3l.herokuapp.com/doc/user-guide/glossary/dvc-project)

jeremydesroches commented:
> Question: do docs in basic-concepts md files support the `description` field in Frontmatter already?

Yes @jorgeorpinel , I've verified that both meta title and meta description work when added to the glossary item frontmatter after @rogermparent's latest changes. So there isn't a problem with trying to address both with the same content, and we won't need to extract the first paragraph either.

jeremydesroches commented:
> I've pushed some commits that that re-enable blog page creation and change sidebar.json to point to the glossary directory to show off docs pages in glossary. I'm certain this is a kludgy representation that will need to be changed, but it gets the idea across.

This is great @rogermparent, thanks! I've merged the concepts and glossary in the latest push. I think this is going to work well... a lot of flexibility between the frontmatter and navigation so we can choose what's needed for each concept.

jorgeorpinel commented on content/docs/sidebar.json:
```diff
@@ -87,6 +87,19 @@
         "source": "what-is-dvc.md"
       },
+      {
+        "label": "Concepts",
```
Will commit this for now:
```suggestion
        "label": "Basic Concepts",
```

jorgeorpinel commented on content/docs/sidebar.json:
```diff
@@ -87,6 +87,19 @@
+          "dvc-metafiles",
+          "dvc-cache",
+          "data-pipelines",
+          "metrics-and-plots",
+          "remote-storage"
```
```suggestion
          "remote-storage",
          "metrics-and-plots"
```

jorgeorpinel commented on content/docs/user-guide/glossary/dvc-cache.md:
```diff
@@ -0,0 +1,22 @@
+tooltip: >-
+  The DVC cache is a hidden storage (by default located in the `.dvc/cache`
+  directory) for files that are tracked by DVC, and their different versions.
+  Learn more about it's
+  [structure](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory).Learn
+  more about the [concept of cache](/doc/user-guide/glossary/dvc-cache) in DVC.
```
> Learn more about the concept of cache

Prob no need to use the term "concept". These links could even be more casual e.g. link "DVC cache" at the beginning of the p (this way you also avoid repeating "Learn more about").

jeremydesroches replied:
OK, updated this and also tried something similar for workspace.

jorgeorpinel replied:
Thanks.

jorgeorpinel commented on content/docs/user-guide/glossary/workspace.md:
```diff
@@ -0,0 +1,19 @@
+tooltip: >-
+  Directory containing all your project files e.g. raw datasets, source code, ML
+  models, etc. Typically, it's also a Git repository. It will contain your DVC
+  project. Learn more about the [workspace
+  concept](/doc/user-guide/glossary/workspace) in DVC.
```
Note that links should be the same line, I think. At least for regular md content, Prettier enforces that (but it doesn't understand that this is also md so ignores it).

jorgeorpinel replied:
> For the links I think it will come down to which items have their own page (basic concepts) and which ones remain as glossary items

Agree, BTW. But don't they all have concepts pages as of now?

jeremydesroches replied:
> it doesn't understand that this is also md so ignores it

Thanks for pointing this out. Tried the shorter version of linking like you suggested with cache here.

> But don't they all have concepts pages as of now?

Technically all the items in the /glossary/ folder have pages and can be linked to. What I meant is some of those items are listed under Basic Concepts in nav and some still only appear as tooltips (and won't have links to the concept page unless we add it).

jorgeorpinel replied:
> some of those items are listed under Basic Concepts in nav and some still only appear as tooltips

Oh like "dependency", yeah OK gotcha. So yes, let's link from the tooltip to the page of all the ones we are actually editing in this PR. Thanks!

jeremydesroches replied:
> let's link from the tooltip to the page of all the ones we are actually editing here

OK, I'll use that pattern. Added placeholder links to each for now.

jorgeorpinel commented:
> both meta title and meta description work...

I see we're not yet using the `description` field in any doc though @jeremydesroches. Planning to? We can (probably should) definitely leave that for a separate PR, although that's kind a major reason for you to have taken this issue on, I think 😬. So if you want to focus more on that, I can definitely help with finalizing the longer concept texts that you've already centralized from their previous locations.

jorgeorpinel commented:
> and we won't need to extract the first paragraph

But if I understand the current approach correctly, we will now end up with 3 sources of text for the same concepts: 1 SEO `description` field, 2 `tooltip` field, and 3 long-form md content for the actual rendered doc. I can see 1 being different enough from the others, but 2 and 3 will overlap a lot. So the idea of extracting the 1st paragraph instead of the `tooltip` field is still appealing to me — if I'm understanding this correctly.

Cc @rogermparent sorry to keep picking on you: do you think that ^ would be doable (instead of the `tooltip` field)? Hacky? Just to know whether it's even an option for now. Thanks

> we can have glossary entries fall back to the body if that field is not present

BTW (also Roger) I'm not sure. Maybe we want to have concepts pages that aren't also tooltips. For now all the concepts will have a `tooltip` text, so it's not a big deal anyway.

jorgeorpinel commented on content/docs/user-guide/glossary/dvc-metafiles.md:
```diff
@@ -0,0 +1,32 @@
+---
+name: 'DVC Metafiles'
+match: ['DVC files', 'files', 'directories']
+tooltip: 'DVC metafiles tooltip...'
+---
+
+# DVC Metafiles
```
Yeah now that I see this I think it's best to keep this here and only leave DVC internals in https://dvc.org/doc/user-guide/dvc-files-and-directories which has gotten too long. But I need @shcheklein to confirm on this plan please, since that guide is probably the most important one we have so far.

p.s. if we keep it we can also mention/reiterate on the (versioning through) codification principle here, which will be explained in the new Versioning use case (#1747) — so no need for that concept stand-alone.

jeremydesroches replied:
> I think it's best to keep this here and only leave DVC internals

Agree on that. Will skip over metafiles for now while we wait for confirmation. In latest deploys you can see how it's looking with cache information extracted already.

> if we keep it we can also mention/reiterate on the (versioning through) codification principle here ... no need for that concept stand-alone

Makes sense. A single source keeps things simple.

jorgeorpinel replied:
> I think it's best to keep this here and only leave DVC internals in https://dvc.org/doc/user-guide/dvc-files-and-directories which has gotten too long

Ping @shcheklein WDYT?

jorgeorpinel replied:
Hi! Update on this one: I did chat with Ivan and we (sort of) agreed to go this way, but will probably change the name to just **DVC Files** here (it may become a section with several pages in the future even), and **DVC Internals** for the guide. No need to do that in this PR; this is just meant as a confirmation.

jorgeorpinel replied:
And another update on this: actually it's uncertain whether moving the DVC Files specs from the current guide to the concepts section will be an improvement. And since we want concept pages to be brief, after all this is probably not a good idea. If it has already been done I'll take care of figuring it out, otherwise let's not do this now. Thanks

rogermparent commented:
> But if I understand the current approach correctly, we will now end up with 3 sources of text for the same concepts: 1 SEO `description` field, 2 `tooltip` field, and 3 long-form md content for the actual rendered doc. I can see 1 being different enough from the others, but 2 and 3 will overlap a lot. So the idea of extracting the 1st paragraph instead of the `tooltip` field is still appealing to me — if I'm understanding this correctly.
>
> Cc @rogermparent sorry to keep picking on you: do you think that ^ would be doable (instead of the `tooltip` field)? Hacky? Just to know whether it's even an option for now. Thanks

No worries! These kinds of changes are what I'm specialized for.

While it's theoretically doable, I think it would end up a little hacky and the mechanics behind it would cause issues that would require many tooltips to be overridden to be usable anyway. The three fields in question: **SEO description**, **Tooltip**, and **body** (or rather its first paragraph) are only near identical right now because two of the fields didn't exist before. If I understand the SEO discussion from where the description field was added, the SEO description field will often not even be populated to leverage search engines' context-driven automatic descriptions.

Considering Tooltips are very likely to be small, keeping them as a parallel text to the body isn't much more overhead and avoids the considerations of "what is a paragraph" and possible oddities of using an arbitrary slice of body text as a tooltip.

> BTW (also Roger) I'm not sure. Maybe we want to have concepts pages that aren't also tooltips. For now all the concepts will have a `tooltip` text, so it's not a big deal anyway.

I think this is a fine idea, if nothing else to allow for the option.

All this considered, the route I'd suggest is: 

- We keep the `tooltip` and `match` fields as the only source for tooltip text (maybe changed into a sort of `{tooltip: {body, match}}` structure in the frontmatter)
- Trim tooltips to the minimal necessary text to be useful while delegating all detail to the page's body. 
- Add a link to the source page (if it exists) into the tooltip component such that any user who needs more info than a concept's tooltip provides can easily go to its main page.
- Disable tooltip creation for all pages with `tooltip` and/or `match` specified.

As always, I'm open to discussion or suggestions on this plan.

jeremydesroches commented:
> I see we're not yet using the `description` field in any doc though @jeremydesroches. Planning to? 

Yes, I plan to add meta descriptions. This could be a good place to use a pattern for in the meta title too. For example, "Concept: Data Pipelines" and the full title would become "Concept: Data Pipelines | Data Version Control · DVC".

> I can definitely help with finalizing the longer concept texts that you've already centralized from their previous locations.

That would be great. I started with `cache` because the information to extract was already pretty well-defined. I'll continue with the rest like that and then work on refining and adding to the extracted parts. We'll also have quite a few links to fix/check.

jorgeorpinel commented:
Ah, sorry for the delay here. Somehow missed all the updates until now...

@rogermparent 

> it would end up a little hacky
> only near identical right now because two of the fields didn't exist before

OK I will think we will have considerable duplicity to manage but let's iterate on this later if it's a problem, since using the 1st p is not ideal implementation-wise.

> the possibility of allowing tooltip on any docs page is doable

Interesting... Will keep that in mind, thanks.

> Disable tooltip creation for all pages with tooltip and/or match specified.

I didn't get this part. Why would we disable all tooltips? Did you mean "without"? 🙂 

@jeremydesroches 

> could be a good place to use a pattern for in the meta title too. For example, "Concept: Data Pipelines

No need for that pattern for concepts, I think. Doesn't seem to add anything IMO.

And OK, I'll take over the content, please finalize the SEO titles, descriptions, and also help me by listing which key words we're going for (so I review them and incorporate them to the content). If you can also check all the links and possibly redirects that would also be great.

jorgeorpinel commented on content/docs/command-reference/cache/dir.md:
```diff
@@ -18,7 +18,7 @@ positional arguments:
 
 Helper to set the `cache.dir` configuration option. (See
-[cache directory](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory).)
+[cache directory](/doc/user-guide/glossary/dvc-cache#structure-of-the-cache-directory).)
```
Can anchored links have their own search history with Google @jeremydesroches ?

I'm thinking whether and how to setup a redirect for this change... Seems tricky since the old content will be split into a new page.

jeremydesroches replied:
> Can anchored links have their own search history with Google @jeremydesroches ?

No, they can't. Depending on content and query an #id anchor and subtext might be used for a snippet, but that's true for all structured content.

So we don't need a redirect on the #id anchors. Only for any base URLs that we remove or change. (Plus it doesn't really work like that... would redirect all `.../files-and-directories` links).

jorgeorpinel replied:
> So we don't need a redirect

jorgeorpinel replied:
> it doesn't really work like that

True...

> So we don't need a redirect

OK! But how do we best dodge losing traffic for removing half of the content from there? I guess adding a prominent link from the guide to the new concepts page? Could be removed later, when someone remembers.

jeremydesroches replied:
> how do we best dodge losing ranking/traffic for moving half of the content out of the guide?

If the content were totally removed there could be an issue. In this case, content updates are encouraged and moving it to another location on the site is not likely to be penalized. The search engine will notice and index the new content. It may even notice the similarity (a black box to us, though). 

When it is reindexed at the new location it should retain its original "value" as a starting point (same content). Assuming our reorganization and addition of concepts is an improvement (i.e., better content structure and engagement) then it will be able to grow on that basis.

>  I guess adding a prominent link from there to the new concepts page?

Yes, exactly, and any link will do in terms of making sure the search engine finds it and reindexes. There will be multiple links, it's in the nav, and it will show up in the sitemap, which is indexed often for DVC.org.

jorgeorpinel replied:
Sounds good. In any case , per my recent https://github.com/iterative/dvc.org/pull/1944#discussion_r536392345, let's not move the details about the file formats to the concepts section (if it hasn't been done).

jorgeorpinel replied:
> let's not move the details about the file formats to the concepts section

I see we haven't yet. OK. Resolving this.

jorgeorpinel commented on content/docs/command-reference/cache/index.md:
```diff
@@ -15,11 +15,10 @@ positional arguments:
-the <abbr>workspace</abbr> are links\* to (or copies of) the ones in cache.
-Learn more about it's
-[structure](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory).
+The <abbr>DVC cache</abbr> is where your data files, models, etc. (anything you
+want to version with DVC) are actually stored. The data files and directories
+visible in the <abbr>workspace</abbr> are links\* to (or copies of) the ones in
+cache.
```
OK, since we moved the concept out of the cmd refs, let's just focus on what the `cache` command (group) is for here please. And link to the concept page

jorgeorpinel replied:
> let's just focus on what the cache command (group) is for

p.s. like it was done with `dag` 👍

jorgeorpinel replied:
ACTUALLY. We may be able to skip the `cache` index completely and have it jump directly to `cache dir` when you click on this...

jorgeorpinel replied:
UPDATE: I've deleted this file and changed the source to `false` in sidebar.json so the nav entry skips to the first child (`dir`). @jeremydesroches please do the same for Parameters once extracted into its concept page.

jeremydesroches replied:
> please do the same for Parameters once extracted into its concept page.

OK, done for Parameters. I also pulled the example from the index there because it was extensive and unique (I think). Good question for the other concepts too: do examples belong or should they remain in cmd ref?

jorgeorpinel replied:
> pulled the example from the (params) index there because it was extensive and unique

Good thinking. Always best to err on the side of caution. I'll finalize the concept pages contents when the time comes.

> Good question for the other concepts too: do examples belong or should they remain in cmd ref?

We want to keep examples in cmd refs. Please don't move those (general rule), just conceptual explanations. Thanks

jorgeorpinel commented on content/docs/user-guide/dvc-files-and-directories.md:
```diff
@@ -239,7 +239,7 @@ Full <abbr>parameters</abbr> (key and value) are listed separately under
 
 - `.dvc/cache`: The <abbr>cache</abbr> directory will store your data in a
-  special [structure](#structure-of-the-cache-directory). The data files and
+  special [structure](/doc/user-guide/glossary/dvc-cache). The data files and
```
Good! There's another 10+ instances of this link/anchor throughout docs though... Let's grep em all

jeremydesroches replied:
OK, done. Where it was possible, tried to preserve style by retaining the anchor link.

jorgeorpinel replied:
All previous links to the anchor should still use the anchor in the new location. Just linking to the cache concept page could be redundant if the doc already has a cache tooltip, for example. Let's assume link to `#structure-of-the-cache-directory` for a good reason, thanks.

jeremydesroches replied:
> All previous links to the anchor should still use the anchor in the new location.

OK, revisited and checked these again... all with original anchor should have it now.

jorgeorpinel commented on content/docs/command-reference/config.md:
```diff
@@ -127,8 +127,8 @@ remote. See `dvc remote` for more information.
 different versions. (See `dvc cache` and
-[DVC Files and Directories](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory)
-for more details.) This section contains the following options:
+[DVC Files and Directories](/doc/user-guide/dvc-files-and-directories) for more
```
Hmmm, I think this should link to /doc/user-guide/glossary/dvc-cache#structure-of-the-cache-directory now, no?

jeremydesroches replied:
Yep, fixed.

jorgeorpinel commented on content/docs/sidebar.json:
```diff
@@ -87,6 +87,19 @@
         "source": "what-is-dvc.md"
       },
+      {
+        "label": "Basic Concepts",
+        "slug": "glossary",
```
BTW. How did we end up with a mix of `glossary` in the URLs but "Basic Concepts" in the nav and title? 😋 

Let's use `concepts` in the slug and URL paths too. We just know this is where we have our glossary (tooltips) internally 🙂 

> Sry, will imply lots of link changes again, prob.

jeremydesroches replied:
Haha, we went back and forth on it. Fixing this first to minimize link changes on the rest.

jorgeorpinel commented on content/docs/user-guide/glossary/data-pipelines.md:
```diff
@@ -0,0 +1,33 @@
+---
+name: 'Data Pipelines'
+match: ['data pipeline', 'pipeline', 'pipelines']
+tooltip: >-
+  In DVC, [data pipeline](/doc/user-guide/glossary/data-pipelines) stages and
+  commands, inputs, outputs, interdependencies, and results (intermediate or
+  final) are specified in `dvc.yaml`, which can be written manually or built
+  using the helper command `dvc run`. This allows DVC to restore one or more
+  pipelines later (see `dvc repro`).
```
Too long/complicated, I think. Probably better to switch with the current 1st paragraph of the body for this one and work it from there. It's best to actually run the dev env and see what the tooltip looks like:

![image](https://user-images.githubusercontent.com/1477535/100529535-38af1180-31ae-11eb-951d-c71f27e08878.png)

jeremydesroches replied:
> Too long/complicated, I think. Probably better to switch with the current 1st paragraph of the body for this one and work it from there.

OK, agreed. Replaced with p1 edited to remove extra links. Also derived the meta description from that (removed markup too).

> It's best to actually run the dev env and see what the tooltip looks like

Yep, I've been doing that. Is this "cleaner" style with only a single link a good pattern to follow?

jorgeorpinel replied:
Thanks.

> Is this "cleaner" style with only a single link a good pattern to follow?

Yes.

jorgeorpinel commented on content/docs/command-reference/dag.md:
```diff
@@ -15,25 +15,8 @@ positional arguments:
-
-`dvc dag` command displays the stages of a pipeline up to the target stage. If
-`target` is omitted, it will show the full project DAG.
+The `dvc dag` command displays the stages of a <abbr>data pipeline</abbr> up to
+the target stage. If `target` is omitted, it will show the full project DAG.
```
We tend to go straight to the verb in these intros:
```suggestion
Displays the stages of a <abbr>data pipeline</abbr> up to the target stage. If
`target` is omitted, it will show the full project DAG.
```
(committing)

jorgeorpinel commented:
I just realized we're missing the Parameters concept, which shouold be extracted from the cmd ref index currently in https://dvc.org/doc/command-reference/params. Added a check box to the PR desc.

jorgeorpinel commented on content/docs/command-reference/cache/index.md:
```diff
@@ -1,37 +0,0 @@
-# cache
-
-Contains a helper command to set the <abbr>cache</abbr> directory location:
```
BTW another kind of link to review is those done automatically with the `` `dvc cache` `` text in other md files. Those used to link here for an explanation of what the cache is, but should be regular links to the concept now, instead (same for `dvc params`). Thanks

jorgeorpinel replied:
\* same for SOME INSTANCES of `dvc params` — need to check the context.

jeremydesroches replied:
Nice... OK, fixed one `dvc cache` link in the docs. Should I fix the one in this blog post too?:

> /blog/2020-04-16-april-20-community-gems.md

jeremydesroches replied:
> * same for SOME INSTANCES of dvc params — need to check the context.

OK, got these too. Didn't change the `dvc params diff` instances.

jorgeorpinel replied:
> Should I fix the one in this blog post too?:
/blog/2020-04-16-april-20-community-gems.md

I don't see any `dvc cache` link in https://dvc.org/blog/april-20-community-gems

jorgeorpinel replied:
Links will be tricky. This is also why we need to split this PR into many so we can take a look at the details concept by concept. Resolvin this one for nwo.

jorgeorpinel commented:
Also, this PR is too big 😢 . We should try to address the comments so far and then break it up somehow... Maybe a first one enabling the concepts feature, and then a PR for each concept.

jorgeorpinel commented on content/docs/command-reference/metrics/index.md:
```diff
@@ -63,10 +33,8 @@ stages:
+Note that metrics files are normally committed with Git (that's what
+`cache: false` above is for). See `dvc.yaml` for more information.
 
 ### Supported file formats
 
```
This should be moved to the concepts page.

Cmd ref indices are all going to end up being pretty small after this change. Maybe even remove them all (but for now this one can stay).

jeremydesroches replied:
> This should be moved to the concepts page.

OK, done.

> Maybe even remove them all.

Yes, would also help with SEO on two fronts: 1. removing duplicate/similar content, 2. expanded/improved content on remaining pages

jorgeorpinel replied:
Looks good, thanks. And yes, I'll prob remove the indices as we work on each concept in it's own PR.

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -17,37 +18,17 @@ positional arguments:
-This type of metrics files are created by users, or generated by user data
-processing code, and can be defined in `dvc.yaml` (`plots` field) for tracking
-(optional).
+### Supported file formats
 
 DVC generates plots as HTML files that can be open with a web browser. These
 HTML files use [Vega-Lite](https://vega.github.io/vega-lite/). Vega is a
```
Same

jeremydesroches replied:
Done

jorgeorpinel commented on content/docs/command-reference/plots/show.md:
```diff
@@ -174,10 +174,60 @@ file:///Users/usr/src/plots/logs.csv.json
-    {
-        "accuracy": "0.9418667",
-    ...
+  "$schema": "https://vega.github.io/schema/vega-lite/v4.json",
+  "data": {
+    "values": [{ "accuracy": "0.9418667", ... }]
+  }
+}
```
No need to change this.

jeremydesroches replied:
OK, reverted. (Was only to fix the sea of red in editor from the open bracket, but now the examples that followed are reverted too.)

jorgeorpinel replied:
Sounds like an editor issue.

jorgeorpinel commented on content/docs/command-reference/plots/show.md:
````diff
@@ -174,10 +174,60 @@ file:///Users/usr/src/plots/logs.csv.json
+}
+```
+
+## Example: Smooth plot
+
+In some cases we would like to smooth our plot. In this example we will use a
+plot with 1000 data points:
+
+```dvc
````
Hmmmm. Where does this come from @jeremydesroches? Seems useful but totally out of scope for this PR?

jorgeorpinel replied:
Ooooh I see it was moved here from the index. Yeah that's not necessary, we have some mixed examples in cmd indices some times. Let's put all those back please 🙂

jeremydesroches replied:
OK, reverted.

jorgeorpinel replied:
Thanks

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -154,106 +159,3 @@ header (first row) are equivalent to field names.
-
-- `-v`, `--verbose` - displays detailed tracing information.
-
-## Example: Tabular data
-
-We'll use tabular metrics file `logs.csv` for this example:
```
Same. No reason to remove these.

jeremydesroches replied:
OK, reverted.

jorgeorpinel commented on content/docs/command-reference/plots/index.md:
```diff
@@ -154,106 +159,3 @@ header (first row) are equivalent to field names.
-
-![](/img/plots_show_smooth.svg)
-
-## Example: Confusion matrix
-
-We'll use `classes.csv` for this example:
```
Same

jeremydesroches replied:
OK, reverted.

jorgeorpinel reviewed (changes requested):
I've reviewed everything in detail except the contents of the concept pages. Let's address all the feedback so far and we'll take it from there. As mentioned recently, this PR will have to be split into several after that. Thanks

jeremydesroches commented:
> I've reviewed everything in detail except the contents of the concept pages. Let's address all the feedback so far and we'll take it from there. As mentioned recently, this PR will have to be split into several after that.

Thanks for the detailed review @jorgeorpinel. I believe I've addressed all of the feedback so far. I've just added meta descriptions for the concepts as well. Will follow-up tomorrow with the keywords for each. Also will add a meta description and keyword for `dvc-metafiles` once that approach is confirmed.

> Also, this PR is too big 😢 . We should try to address the comments so far and then break it up somehow... Maybe a first one enabling the concepts feature, and then a PR for each concept.

Yep, makes sense to enable with parity to existing glossary, then enable each concept as completed. What's the best way (technically) to do that from here with all the WIP? Hoping for a GitHub pro-tip... 😉

jorgeorpinel commented on content/docs/command-reference/config.md:
```diff
@@ -126,8 +126,8 @@ remote. See `dvc remote` for more information.
-different versions. (See `dvc cache` and
-[DVC Files and Directories](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory)
+different versions. (See
+[DVC cache](/doc/user-guide/concepts/dvc-cache#structure-of-the-cache-directory)
 for more details.) This section contains the following options:
```
Actually given the context in this case no anchor is needed:
```suggestion
[DVC cache](/doc/user-guide/concepts/dvc-cache) for more details.) This section
contains the following options:
```

jorgeorpinel commented:
> enable with parity to existing glossary, then enable each concept as completed. What's the best way (technically) to do that from here

1. First I'd merge the latest master to this branch in case there are any merge conflicts (not the case rn).
2. Create a new branch from (latest) master, move there, and `git checkout ug-basic-concepts -- [file] [file] ...` the files for the first PR and open a PR from that branch.
3. Same for the other PRs one by one, basically. I guess I can finish up the PR split but at least having the base one and a first concept from you would be nice.

Needless to say this PR would be closed.

For now I'll review the SEP meta/desc fields here.

jorgeorpinel commented:
p.s. perhaps the first concept PR (after the PR that enables this functionality) can be for the DVC Cache, so that most of the cmd refs are included there, as they mostly have the structure-of-cache link change.

jorgeorpinel commented on content/docs/command-reference/metrics/index.md:
````diff
@@ -63,33 +33,8 @@ stages:
           cache: false
 ```
 
-> `cache: false` above specifies that `summary.json` is not tracked or
-> <abbr>cached</abbr> by DVC (`-M` option of `dvc run`). These metrics files are
-> normally committed with Git instead. See `dvc.yaml` for more information on
-> the file format above.
````
Looks like this note should stay? It refers to the code block above, which is left in the ref.

jorgeorpinel replied:
Ah never mind I see we rewrote this as a regular p.

jorgeorpinel commented on content/docs/command-reference/params/diff.md:
```diff
@@ -1,7 +1,7 @@
 
-Show changes in [parameter dependencies](/doc/command-reference/params) between
-commits in the <abbr>DVC repository</abbr>, or between a commit and the
+Show changes in [parameter dependencies](/doc/user-guide/concepts/parameters)
```
We should use tooltips instead of links to `/doc/user-guide/concepts/parameters` (in most places, depends on the context) now.

jeremydesroches replied:
OK, good idea. Updated these ones (even though I know we'll be splitting this up).

jorgeorpinel replied:
> Updated these ones (even though I know we'll be splitting this up

Any progress here will be preserved when I can finally split this so thanks.

jorgeorpinel commented on content/docs/command-reference/params/diff.md:
```diff
@@ -23,7 +23,7 @@ in the repository history. Requires that Git is being used to version the
 
 > Parameter dependencies are defined with the `-p` option in `dvc run`. See also
-> `dvc params`.
+> [parameters](/doc/user-guide/concepts/parameters).
```
Same

jeremydesroches replied:
Done

jorgeorpinel commented on content/docs/command-reference/params/diff.md:
```diff
@@ -52,7 +52,7 @@ itself does not ascribe any specific meaning for these values.
 
 Let's create a simple YAML parameters file named `params.yaml` (default params
-file name, see `dvc params` to learn more):
+file name, see [parameters](/doc/user-guide/concepts/parameters) to learn more):
```
Same

jeremydesroches replied:
Done

jorgeorpinel commented on content/docs/command-reference/push.md:
```diff
@@ -185,7 +185,7 @@ Finally, we used `dvc status` to double check that all data had been uploaded.
 
 Let's take a detailed look at what happens to the
-[cache directory](/doc/user-guide/dvc-files-and-directories#structure-of-the-cache-directory)
+[cache directory](/doc/user-guide/concepts/dvc-cache#structure-of-the-cache-directory)
```
Kind of unrelated to the PR but we probably don't need this link since it's also in a block quote by the end of the example. But maybe could be a tooltip? If there aren't other tooltips to this concept nearby already.

jeremydesroches replied:
Agreed. Added the tooltip here and added `cache directory` to the available matches in the concept.

There is another `cache` tooltip in the next paragraph. Too many?

jorgeorpinel replied:
I'd just use `<abbr>cache</abbr> directory` and avoid the extra glossary match entry.

> There is another cache tooltip in the next paragraph. Too many?

Yep, I'd remove that later one. Ugh managing `<abbr>` tags can be painful.

jorgeorpinel commented on content/docs/command-reference/run.md:
```diff
@@ -193,8 +193,8 @@ $ dvc run -n my_stage './my_script.sh $MYENVVAR'
   `params.yaml`, but this can be redefined with a prefix in the argument sent to
-  this option, e.g. `-p parse_params.yaml:threshold`. See `dvc params` to learn
-  more about parameters.
+  this option, e.g. `-p parse_params.yaml:threshold`. See
+  [parameters](/doc/user-guide/concepts/parameters) to learn more.
```
> Right, sometimes an auto-link to the cmd ref may be most appropriate.

jorgeorpinel replied:
But actually in this case the default params file name info has been moved to the concepts page so a link to concepts page would be better here... (for now, not sure where I'll keep all that detailed params behavior info).

jorgeorpinel replied:
And there's another 2 instances of this change down this dic.

jorgeorpinel replied:
doc

jeremydesroches replied:
Replaced these three with tooltips, plus another one in the description. (There are also similar references for Metrics and Plots that may need similar treatment.)

jorgeorpinel commented on content/docs/user-guide/concepts/data-pipelines.md:
```diff
@@ -0,0 +1,37 @@
+  of data processing stages (for example, console commands that take an input
+  and produce an output). A pipeline may produce intermediate data, and has a
+  final result.
+description: >-
```
What about SEO titles?

jeremydesroches replied:
> What about SEO titles?

In our previous discussion we discussed adding the "Concept:" prefix and the decision was to leave it out. That would be the only change I'd make so otherwise the generated titles are OK.

jorgeorpinel replied:
> In our previous discussion we discussed adding the "Concept:" prefix and the decision was to leave it out. 

Yes. But I thought we add `titles` everywhere as a "strong suggestion" to Google on what to use as title, since it tends to decide the title on it's own anyway 😆 — you call.

jeremydesroches replied:
Our discussions on metadata have mostly been about the `description`, which does need to be specified because otherwise it defaults to the same copy for every URL.

For the `title` though, if we don't add it in frontmatter it's generated from the label in `sidebar.json`. So we are giving Google a "strong suggestion" for the `title` on each doc URL already. 

My thinking was to only add the title in frontmatter if it needs to be different from the navigation label (to add a prefix, for example). Along the lines of "don't repeat yourself" — but I don't have a problem with being explicit about it either. WDYT?

jorgeorpinel replied:
> if we don't add it in frontmatter it's generated from the label in sidebar.json

Ah OK then, I didn't remember that. Thanks

jorgeorpinel commented on content/docs/user-guide/concepts/data-pipelines.md:
```diff
@@ -0,0 +1,37 @@
+name: 'Data Pipelines'
+match: ['data pipeline', 'pipeline', 'pipelines']
+tooltip: >-
+  In DVC, a [data pipeline](/doc/user-guide/concepts/data-pipelines) is a series
+  of data processing stages (for example, console commands that take an input
+  and produce an output). A pipeline may produce intermediate data, and has a
+  final result.
```
```suggestion
  A [data pipeline](/doc/user-guide/concepts/data-pipelines) is a series of data
  processing stages, chained by their outputs and inputs. They use some initial
  data, may produce intermediate artifacts, and reach a final result.
```

jorgeorpinel commented on content/docs/user-guide/concepts/data-pipelines.md:
```diff
@@ -0,0 +1,36 @@
+  processing stages, chained by their outputs and inputs. They use some initial
+  data, may produce intermediate artifacts, and reach a final result.
+description: >-
+  In DVC, a data pipeline is a series of data processing stages. A pipeline may
+  produce intermediate data, and has a final result.
```
- No need for "In DVC" I think
- Maybe use "They use some initial data, may produce intermediate artifacts, and reach a final result" (text I used in the tooltip)
- What key terms are we emphasizing?
- Has there been a SEO research on existing terms (per the GSoD project proposal)? E.g. I suspect "workfow" could be a good term to incorporate here (not just in the SEO desc but in the content).

jeremydesroches replied:
Updated the description with your suggestions. 

The existing terms to focus on are listed in the HTML comment, and I've added some more around "workflow" that I verified in search console (good idea).

Terms in the description have no direct impact on ranking, but the clickthrough ratio does. So for the description we should be doing our best to improve clickthrough (looking at terms with most potential). Based on DVC traffic, it's for sure "data pipelines", and the ML, data science, *ops terms listed are speculative based on related terms and searches.

Terms in the content do have a direct impact on the ranking, so that's where working in the keywords is most worthwhile.

jeremydesroches replied:
> So for the description we should be doing our best to improve clickthrough (looking at terms with most potential). 

An idea for how to improve CTR: Since this is a concept the first sentence of the description is good (definition of a pipeline). For the second sentence (~100 characters), we could use one or two key terms to describe uses for pipelines in DVC. Something like "Use DVC to build a machine learning pipeline or data science workflow."

> Terms in the content do have a direct impact on the ranking, so that's where working in the keywords is most worthwhile.

To further highlight the difference in approach: "workflow" comes up in a few different cases, so to improve rank beyond adding keywords, we might adjust or create an example to describe "building a machine learning (data science) workflow using pipelines in DVC". The better we can answer the query the more likely it will be rewarded. (Keeping in mind that we have to choose which queries are worth the effort.)

jorgeorpinel replied:
Awesome, thanks for the tips!

jorgeorpinel reviewed (changes requested):
Looks like SEO metadata is missing from the dependency, dvc-cache, dvc-metafiles, output concepts (there are no changes to those files). All the frontmatter in fact (inc. draft tooltip)

jorgeorpinel commented on content/docs/user-guide/dvc-files-and-directories.md:
```diff
@@ -250,7 +250,7 @@ Full <abbr>parameters</abbr> (key and value) are listed separately under
 
 - `.dvc/cache`: The <abbr>cache</abbr> directory will store your data in a
-  special [structure](#structure-of-the-cache-directory). The data files and
+  special [structure](/doc/user-guide/concepts/dvc-cache). The data files and
```
Missing anchor 🙂

jeremydesroches replied:
Fixed. Thanks!

jorgeorpinel commented on content/docs/user-guide/large-dataset-optimization.md:
```diff
@@ -4,7 +4,7 @@ In order to track the data files and directories added with `dvc add` or
 <abbr>project</abbr>'s cache is the hidden storage (by default located in
 `.dvc/cache`) for files that are tracked by DVC, and their different versions.
-(See `dvc cache` and
+(See [DVC cache](/doc/user-guide/concepts/dvc-cache) and
```
```suggestion
(See [DVC Cache](/doc/user-guide/concepts/dvc-cache) and
```

jorgeorpinel commented:
> > What's the best way (technically) to do that from here
>
> First I'd merge the latest master...

You know what @jeremydesroches I can split this PR when ready. Let's focus on addressing feedback from your changes so far and especially on the SEO stuff. Ad each part seems ready I'll open parallel PRs with the actual intention to close them and ask Ivan to review those. Thanks

jorgeorpinel commented on content/docs/user-guide/concepts/dvc-cache.md:
```diff
@@ -0,0 +1,87 @@
+  The [DVC cache](/doc/user-guide/concepts/dvc-cache) is a hidden storage (by
+  default located in the `.dvc/cache` directory) for files that are tracked by
+  DVC, and their different versions.
+description: >-
+  The DVC cache adds a layer of indirection between code and data to efficiently
+  version large datasets, data science features, and machine learning models.
```
Same general Qs as with data-pipelines: what terms did the SEO/search history research throw for this one?

So far all the descriptions seem like excerpts from previous content we had. I was looking for some new stuff based on SEO research.

Can we list the terms/goals for each concept's SEO desc? Thanks

jeremydesroches replied:
> Looks like SEO metadata is missing from the dependency, dvc-cache, dvc-metafiles, output concepts (there are no changes to those files). All the frontmatter in fact (inc. draft tooltip)

Not following you here... dependency and output are glossary terms but aren't in the list of concepts. For dvc-cache, it is a full concept but it does have the frontmatter in the latest. 🤔  So I was only going to add frontmatter to dvc-metafiles. Is that different from what you were thinking @jorgeorpinel ?

jorgeorpinel replied:
Ping @jeremydesroches

jeremydesroches replied:
> I was looking for some new stuff based on SEO research.

> Can we list the terms/goals for each concept's SEO desc?

I added the last half of the description based on the terms, which are listed in the HTML comments.

jorgeorpinel replied:
I see. Alright.

jorgeorpinel commented on content/docs/user-guide/concepts/dvc-metafiles.md:
```diff
@@ -0,0 +1,37 @@
+---
+name: 'DVC Metafiles'
+match: ['DVC files', 'files', 'directories']
+tooltip: >-
+  'DVC [metafiles](/doc/user-guide/concepts/dvc-metafiles) tooltip...'
+---
```
p.s. You can list them as comments in each file change if you'd like @jeremydesroches, seems like the easiest approach to me.

p.p.s. This one (dvc-metafiles) doesn't have SEO metadata yet.

jorgeorpinel replied:
Ping @jeremydesroches

jeremydesroches replied:
Based on your [comment](https://github.com/iterative/dvc.org/pull/1944#discussion_r536392345) way up there ☝️ it seems like you are saying you don't want to move the Files content or create this item at all. Is that right @jorgeorpinel? If I'm mistaken on this, not a problem to research and create the metadata.

Otherwise, I think I've addressed the remaining feedback so we can start splitting up this PR to work on the pieces in order as you suggested.

jorgeorpinel replied:
You're right but now we just need to revert moving content from https://dvc.org/doc/user-guide/dvc-files-and-directories to here. I guess I can just leave it out when I split this PR, so 👍

jeremydesroches commented:
> Looks like SEO metadata is missing from the dependency, dvc-cache, dvc-metafiles, output concepts (there are no changes to those files). All the frontmatter in fact (inc. draft tooltip)

Not following you here... dependency and output are glossary terms but aren't in the list of concepts. For dvc-cache, it is a full concept but it does have the frontmatter in the latest. 🤔  So I was only going to add frontmatter to dvc-metafiles. Is that different from what you were thinking @jorgeorpinel ?

jorgeorpinel commented:
P.s. I pinged you in the few SEO-related questions left here, Jeremy. I can take care of the rest after that.

jorgeorpinel commented:
Ping @jeremydesroches if you no longer have the capacity to finish the SEO-related parts (see conversations above) just lmk please. Thanks.

jeremydesroches commented:
> Ping @jeremydesroches if you no longer have the capacity to finish the SEO-related parts (see conversations above) just lmk please. Thanks.

Sorry for the delay @jorgeorpinel. I'll get back to you on these items and most of the others today.

jorgeorpinel commented:
No worries @jeremydesroches. I'm going to have low capacity for a few weeks anyway but I'll come back to splitting this ASAP. It looks like so far you've completed your contributions so thanks a bunch! unfortunately I have to close this PR but your commits will be preserved in following PRs which I'll link below for the record (later). Closing for now.
